stan/math/prim/fun/gaus_interp.hpp

```diff
@@ -0,0 +1,202 @@
+#ifndef STAN_MATH_PRIM_FUN_GAUS_INTERP
+#define STAN_MATH_PRIM_FUN_GAUS_INTERP
+
+#include <stan/math/prim/fun/conv_gaus_line.hpp>
+#include <stan/math/prim/fun/square.hpp>
+#include <stan/math/prim/err.hpp>
+#include <cmath>
+#include <vector>
+#include <algorithm>
+
+namespace stan {
+namespace math {
+
+// struct for passing precomputation data
+struct gaus_interp_params {
+  std::vector<double> as;
+  std::vector<double> bs;
+  double sig2;
+};
+
+namespace internal {
+
+/*
+find the smallest difference between successive elements in a sorted vector
+*/
+template <typename Tx>
+inline double min_diff(int n, const std::vector<Tx>& xs) {
+  double dmin = value_of(xs[1]) - value_of(xs[0]);
+  for (int i = 1; i < n - 1; i++) {
+    if (value_of(xs[i + 1]) - value_of(xs[i]) < dmin) {
+      dmin = value_of(xs[i + 1]) - value_of(xs[i]);
+    }
+  }
+  return dmin;
+}
+
+}  // namespace internal
+
+/**
+ * Given a set of reference points \f$(xs_i, ys_i)\f$, create a mollifier
+ * that intersects the reference points. This function requires as input
+ * a struct created by the function gaus_interp_precomp. The algorithm
+ * used to create the mollifier is an iterative algorithm that works
+ * as follows. First a linear
+ * interpolation is created through the reference points. Then, the
+ * linear interpolation is convolved with a Gaussian whose width is
+ * proportional the smallest distance between successive points
+ * \f$xs_i\f$ and \f$xs_{i+1}\f$. Since the convolution is unlikely to
+ * intersect the reference points, the y-values of the reference points
+ * are shifted and the process repeats itself until the interpolation
+ * intersects all reference points.
+ *
+ * Note: This interpolation scheme should be used when the function
+ * to be interpolated is well-resolved by the reference points.
+ *
+ * @param xs vector of independent variable of reference points
+ * @param ys vector of dependent variable of reference points
+ * @param params a struct created by gaus_interp_precomp that
+ * @param x the point at which to evaluate the interpolation
+ * @return value of the interpolation at x
+ */
+template <typename Tx>
+inline return_type_t<Tx> gaus_interp(const std::vector<double>& xs,
+                                     const std::vector<double>& ys,
+                                     const gaus_interp_params& params,
+                                     const Tx& x) {
```

Comment:
If a user calls this you'll need a version that takes in params as an `std::vector<double>` since idt we plan to expose the struct `gaus_interp_params` to the language

Also do you want this to work on eigen vectors (Stan's `vector[]` type) as well? We'd need to modify this a bit like 

```suggestion
template <typename VecX, typename VecY, typename Tx, 
  require_all_vector_like_t<VecX, VecY>* = nullptr>
inline return_type_t<VecX, VecY, Tx> gaus_interp(const VecX& xs,
                                     const VecY& ys,
                                     const gaus_interp_params& params,
                                     const Tx& x) {
```

And use some type traits stuff in stan to deal with the two vector types. I can help out with that once the rest of this is tidied up

Reply:
I don't think it was discussed whether this will have to work on Stan's `vector[]` types. I will tidy up the rest and then it'll be good to discuss further.